ex_app/lib/service.py

```diff
@@ -19,11 +18,52 @@ class Application:
 	def __init__(self) -> None:
 		self.hpb_settings = get_hpb_settings()
 		self.spreed_clients: dict[str, SpreedClient] = {}
-		self.spreed_clients_lock = threading.RLock()
-
-	async def transcript_req(self, req: TranscribeRequest) -> None:
-		with self.spreed_clients_lock:
+		self.spreed_clients_lock = asyncio.Lock()
+		self.__task_bin: set[asyncio.Task] = set()
+
+	async def __defer_start_client(self, req: TranscribeRequest) -> None:
+		t = 0
+		# defer starting a new client to give time for cleanup of the previous one
+		# wait up to HPB_SHUTDOWN_TIMEOUT seconds
+		while t <= HPB_SHUTDOWN_TIMEOUT:
+			t += 5
+			await asyncio.sleep(t)
+			async with self.spreed_clients_lock:
```

Comment:
I wonder if we actually need these locks, as asyncio is single-threaded. Thy don't hurt, of course

Reply:
yeah I thought about removing them too but then realised that for places where there are `await`s and it is required to complete all the reads/changes to that resource in order, locks are necessary.
and in places like the one reviewed above, it is good to wait for the lock otherwise the lock in other places won't be any good.

also, spotted a silly mistake, looking up the code again in this context hehe.